02_cancer_type_classification/lasso_range_analysis/nbconverted/compare_optimizers_gene.py

```diff
@@ -0,0 +1,280 @@
+                                                              gene=plot_gene):
+    (gene,
+     cancer_type,
+     seed,
+     lasso_param,
+     coefs_list) = coef_info
+    for fold_no, coefs in enumerate(coefs_list):
+        ll_nz_coefs_df.append(
+            [gene, cancer_type, lasso_param, seed, fold_no, len(coefs)]
+        )
+        
+ll_nz_coefs_df = pd.DataFrame(
+    ll_nz_coefs_df,
+    columns=['gene', 'cancer_type', 'lasso_param', 'seed', 'fold', 'nz_coefs']
+)
+ll_nz_coefs_df.lasso_param = ll_nz_coefs_df.lasso_param.astype(float)
+ll_nz_coefs_df.head()
+
+
+# In[4]:
+
+
+sgd_nz_coefs_df = []
+
+# get coefficient info for training dataset specified above
+for coef_info in au.generate_nonzero_coefficients_lasso_range(sgd_results_dir,
+                                                              gene=plot_gene):
+    (gene,
+     cancer_type,
+     seed,
+     lasso_param,
+     coefs_list) = coef_info
+    for fold_no, coefs in enumerate(coefs_list):
+        sgd_nz_coefs_df.append(
+            [gene, cancer_type, lasso_param, seed, fold_no, len(coefs)]
+        )
+        
+sgd_nz_coefs_df = pd.DataFrame(
+    sgd_nz_coefs_df,
+    columns=['gene', 'cancer_type', 'lasso_param', 'seed', 'fold', 'nz_coefs']
+)
+sgd_nz_coefs_df.lasso_param = sgd_nz_coefs_df.lasso_param.astype(float)
+sgd_nz_coefs_df.head()
+
+
+# In[5]:
+
+
+sns.set({'figure.figsize': (12, 10)})
+fig, axarr = plt.subplots(2, 1)
+
+def color_boxplot_edges(ax):
+    # color the boxplot lines/edges rather than the box fill
+    # this makes it easier to discern colors at the extremes; i.e. very many or few nonzero coefs
+    # https://stackoverflow.com/a/72333641
+    box_patches = [patch for patch in ax.patches if type(patch) == matplotlib.patches.PathPatch]
+    num_patches = len(box_patches)
+    lines_per_boxplot = len(ax.lines) // num_patches
+    for i, patch in enumerate(box_patches):
+        # set the linecolor on the patch to the facecolor, and set the facecolor to None
+        col = patch.get_facecolor()
+        patch.set_edgecolor(col)
+        patch.set_facecolor('None')
+
+        # each box has associated Line2D objects (to make the whiskers, fliers, etc.)
+        # loop over them here, and use the same color as above
+        for line in ax.lines[i * lines_per_boxplot: (i + 1) * lines_per_boxplot]:
+            line.set_color(col)
+            line.set_mfc(col)  # facecolor of fliers
+            line.set_mec(col)  # edgecolor of fliers
+
+    # also fix the legend to color the edges rather than fill
+    for legpatch in ax.legend_.get_patches():
+        col = legpatch.get_facecolor()
+        legpatch.set_edgecolor(col)
+        legpatch.set_facecolor('None')
+
+
+sns.boxplot(
+    data=ll_nz_coefs_df.sort_values(by=['cancer_type', 'lasso_param']),
+    x='cancer_type', y='nz_coefs', hue='lasso_param', ax=axarr[0]
+)
+color_boxplot_edges(axarr[0])
+axarr[0].set_title('liblinear')
+
+sns.boxplot(
```

Comment:
Do you want the legend to be in the middle of the plot? It sometimes overlaps the data, but you can still see the main idea of the plot so your call.